src/compressor/zlib/ZlibCompressor.cc

```diff
@@ -203,7 +205,7 @@ int ZlibCompressor::decompress(bufferlist::const_iterator &p, size_t compressed_
   strm.next_in = Z_NULL;
 
   // choose the variation of compressor
-  ret = inflateInit2(&strm, ZLIB_DEFAULT_WIN_SIZE);
+  ret = inflateInit2(&strm, compressor_message);
```

Comment:
Just realized that this (coupled with setting compressor_message to 0 when decoding legacy headers, e.g. in BlueStore) might cause passing 0 to inflate which isn't correct.
So one should distinguish the following cases when decompressing:
- legacy header , compressor message is uninitialized which effectively is ZLIB_DEFAULT_WIN_SIZE
- new header, compressor message is 0. WIN_SIZE = 0.

I can see some ways to resolve this.
1) make UNINITIALIZED const for compressor_message. And set it to e.g. MAX_INT or any other value presumably not used by any compressor alg?
2) mask 0 at ZLIB level with some other value and treat 0 on input as UNINITIALIZED -> ZLIB_DEFAULT_WIN_SIZE
One will need to do the same for other compressors when start using compressor message.
3) use boost::optional to transport compressor message from upper layers to compress/decompress and do conditional  encoding/decoding (bool flag + optional value) in corresponding headers.
Personally I'd rather prefer 3) as the most straightforward and error-prone

Reply:
good idea,I'll fix it right away